Research Extension

TechCor8 opened the pull request:
Created Research and TechInfo classes.

DanRStevens commented on Source/Research.h:
```diff
@@ -0,0 +1,71 @@
+	tcSpace			= 8, // 8 = Space (spaceport, observatory, launch vehicle, skydock)
+	tcMorale		= 9, // 9 = Population (happiness)
+	tcDisaster		= 10,// 10 = Disaster warning (and defense)
+	tcPopulation	= 11,// 11 = Population (health, growth)
```
Indentation looks off for this line on GitHub. Should perhaps use spaces for justifying comments at the end of lines.

DanRStevens commented on Source/TethysGameEx.cpp:
```diff
@@ -32,7 +32,7 @@ struct OP2Game
 TethysGameEx gGame;
 void *gameObj;
-void *researchObj;
+//void *researchObj;
```
Should just remove this.

DanRStevens commented on Source/TethysGameEx.h:
```diff
@@ -27,7 +27,7 @@ class TethysGameEx : public TethysGame
 extern void *gameObj;
 
-extern void *researchObj; // TODO: move to research file when research functions are written
+//extern void *researchObj; // TODO: move to research file when research functions are written
```
Should just remove this.

DanRStevens commented on Source/Research.cpp:
```diff
@@ -0,0 +1,197 @@
+TechInfo Research::GetTechInfo(int index)
+{
+	TechInfo info;
+	info.internalPtr = (void*)HFLNOTINITED;
```
I think `nullptr` would be more sensible here.

DanRStevens commented on Source/Research.h:
```diff
@@ -0,0 +1,71 @@
+	//int GetNumDependentTech(); // (number of techs that depend on this one) (below is a pointer to an int? list of this many elements)
+	//int* GetDependentTechNum(int index); // (pointer to array of techNums of techs that depend on this one)
+
+	void *internalPtr;
```
Should probably type this as `OP2TechInfo*` to reduce the need for casting.

Or eliminate the proxy struct, as mentioned elsewhere.

TechCor8 replied:
The use of `void* internalPtr` mimics the same use in UnitInfo. Although I agree with changing this to use `OP2TechInfo*`.

DanRStevens replied:
Understood.

I've changed similar code in HFL in the past couple of months. Using the correct type to begin with greatly reduced the need for casting within the library code.

Looks like there was more of it.

DanRStevens commented on Source/Research.cpp:
```diff
@@ -0,0 +1,197 @@
+	if (!isInited)
+		return HFLNOTINITED;
+
+	return (internalPtr != NULL);
```
Newer code is starting to prefer `nullptr`. Older code that is still using `NULL` may trigger compiler warnings in some contexts (if such warnings are enabled).

Brett208 replied:
I agree we want to switch to nullptr. Since nullptr isn't called anywhere else in HFL yet, it doesn't make sense to require it before merging a pull request. Probably better to open an issue and work on converting the code to use nullptr.

TechCor8 replied:
I decided to change this to nullptr. I'm a bit outdated. I haven't used C++ at work since 2009, so I didn't know about it until I saw it used in OP2Utility.

DanRStevens replied:
Brett makes a good point. HFL doesn't yet make use of `nullptr`, though does use `NULL` in a number of places. We'll have to do a post merge pass to update the library.

Looks like this is where we're starting with the modernization.

Brett208 commented:
For HFLNOTINITED, I would have probably designed this library it to use std::exception instead of a return value (although I recognize it was written before exceptions were well established in C++). Since no exceptions are yet thrown in this code base, I would expect any PRs against the code to follow the established convention of using HFLNOTINITED. I suspect code that @Arklon1 and sirbomber use in mission development rely on this behaviour and I'm not sure if we want to change it or not. It would force them to add exception handling to their code.

I really like that HFL provides a layer for developers that is relatively safe and free of memory hacking when used in mission designs. Most people designing missions are not going to have the skills or understanding to complete said activities and would be much more comfortable with a get/set paradigm (this includes me).

As for using pointer casting instead of the boilerplate get /set functions everywhere, I don't have a good understanding of the implementations to really comment on which I would prefer. I wouldn't want to force that discussion on this PR. If this PR follows the established functionality of the library, then it should probably be merged as such. If the pointer casting provides the same ease of use to the end developer as the get/set approach, is less error prone, and requires less lines of code, I don't see why we shouldn't support in the long term.

It does trouble me that we keep redefining all the structures in different libraries. I see the same definitions peppered in mission DLL projects, Outpost2DLL, HFL, OP2Internal, etc. Probably not worth trying to solve here though.

TechCor8 commented:
My opinion on HFLNOTINITED is to make it a separate issue and change all of it to exceptions or asserts at the same time. HFL not being initialized is truly an exceptional case, and I think most, if not all, projects will not need any exception handling.

Why check for exceptions or HFLNOTINITED? If you get that, you forgot to init at start! However, I could be wrong, and people are doing very strange things related to HFL initialization.

DanRStevens commented:
Agreed. We should create a new issue concerning use of exceptions to report errors. Most code won't need to check for exceptions. An error with HFL initialization is a development concern, not a runtime release concern.

As for pointer casting, that's something that should be hidden inside the API. I'm not suggesting pointer casting be pushed out into client code. Just that the API exposed to clients should use real objects rather than proxy objects. Again though, this is perhaps a library wide concern that needs its own issue.

Brett208 commented:
Looks like I did a poor job merging #23 which caused some conflicts in different branches. Should be fixed now though.

DanRStevens commented:
Looks like the merge commit deleted most of this branch. We should probably undo the merge commit. If there are conflicts, perhaps rebase on master.

----

Remaining comments:
- `TechCategory` formatting of enum values (and comments)
- Use proper `OP2TechInfo*` type for `internalPtr`

Brett208 commented:
@DanRStevens I believe when I merged the UnitEx branch, for some reason it brought in all changes from techCor's original post. So those changes were not deleted, but are already incorporated in HFL. 

I don't understand how these PRs were formed. There are no branches on TechCor's fork of the REPO. I think when I used the 'merge pull request' button on GitHub, it just merged the master of his forked branch onto HFL. I don't understand how these PRs were formed without associated branches in the forked repository. Sorry for screwing this up.

TechCor8 commented:
The original pull request was from the fork master, where I had put all the new work in one commit.

To split the work across multiple branches, I checked out the commit before all the new stuff and created a branch "pull-branch" from it. I moved the new work here, but did not commit.

Then I created a request branch from the "pull-branch", moved all my changes to it, committed a subset of the changes. Repeat by creating a new branch from the "pull-branch", moving the remaining changes, committing a subset, etc, until all the request branches were made.

Somehow, the master was merged. I don't know why that would happen. I guess next time I should just blow the whole fork away and start from scratch.

Brett208 commented:
@TechCor8, I'll pay more attention next time before merging. Now that I read through your post, I see the branches you mentioned on your fork. I think maybe if done a bit differently, you might be able to see this PR on your fork. It has been a long time since I have worked with PRs from forks instead of just branches on a repo.

I'm going to go ahead and merge the remainder of this branch. Tonight I'll try to tackle the 2 remaining issues or open issues if they are too long to accomplish.

DanRStevens commented:
Yeah, admittedly most of my experience has been with PR's originating from the same repository. In those cases, I'm able to make my own little fixups to them. With a PR originating from a fork, that's not as easy. It would require adding the fork as a Git remote, fetching the changes, making edits on a local branch, and then pushing changes up and opening a new PR to merge into the branch of the fork. (We don't have automatic write access to the fork, hence the PR to the fork for a PR from the fork).

I was going to make a few minor pre-merge changes to some of these, but got distracted trying to figure out how that's even done with a fork, and never got around to the actual changes.

Anyway, very pleased to see you two getting things sorted out.

Btw, TechCor is now a member of the OPU organization on GitHub, so can now work directly from the master repositories, without needing to fork. (Procedure is still to create the changes on a branch, push to GitHub, and then open a PR to merge them to master. That way code can still be reviewed before merge. The workflow also has the added benefit that any CI builds to validate the changes will be triggered when the branch is pushed to GitHub, or when the PR is opened. Currently though, HFL doesn't have any CI builds).
